File: library/src/main/java/me/proxer/library/api/EnumRetrofitConverterFactory.java
package me.proxer.library.api;

import me.proxer.library.util.ProxerUtils;
import retrofit2.Converter;
import retrofit2.Retrofit;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

/**
 * @author Ruben Gees
 */
final class EnumRetrofitConverterFactory extends Converter.Factory {

    @Override
    public Converter<?, String> stringConverter(final Type type, final Annotation[] annotations,
                                                final Retrofit retrofit) {
        if (((Class<?>) type).isEnum()) {
            return (Converter<Enum<?>, String>) ProxerUtils::getApiEnumName;
        }

        return null;
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/EnumRetrofitConverterFactory.kt
package me.proxer.library.internal.adapter

import me.proxer.library.util.ProxerUtils
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

/**
 * @author Ruben Gees
 */
internal class EnumRetrofitConverterFactory : Converter.Factory() {

    override fun stringConverter(
        type: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit
    ): Converter<Enum<*>, String>? {
        return when {
            (type as Class<*>).isEnum -> Converter { ProxerUtils.getSafeApiEnumName(it) }
            else -> null
        }
    }
}
